serde_json_path_core/src/node.rs

```diff
@@ -209,6 +211,391 @@ impl<'a> NodeList<'a> {
+    ///     .query_located(&value)
+    ///     .iter()
+    ///     .map(|q| (q.location().to_string(), q.node()))
+    ///     .collect();
+    /// assert_eq!(pairs[0], ("$['foo'][0]".to_owned(), &json!("bar")));
+    /// assert_eq!(pairs[1], ("$['foo'][1]".to_owned(), &json!("baz")));
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn iter(&self) -> Iter<'_, LocatedNode<'a>> {
+        self.0.iter()
+    }
+
+    /// Get an iterator over the locations of nodes within a [`LocatedNodeList`]
+    ///
+    /// # Usage
+    /// ```rust
+    /// # use serde_json::json;
+    /// # use serde_json_path::JsonPath;
+    /// # fn main() -> Result<(), serde_json_path::ParseError> {
+    /// let value = json!({"foo": ["bar", "baz"]});
+    /// let path = JsonPath::parse("$.foo.*")?;
+    /// let locations: Vec<String> = path
+    ///     .query_located(&value)
+    ///     .locations()
+    ///     .map(|loc| loc.to_string())
+    ///     .collect();
+    /// assert_eq!(locations, ["$['foo'][0]", "$['foo'][1]"]);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn locations(&self) -> Locations<'_> {
+        Locations { inner: self.iter() }
+    }
+
+    /// Get an iterator over the nodes within a [`LocatedNodeList`]
+    pub fn nodes(&self) -> Nodes<'_> {
+        Nodes { inner: self.iter() }
+    }
+
+    /// Deduplicate a [`LocatedNodeList`] and return the result
+    ///
+    /// See also, [`dedup_in_place`][LocatedNodeList::dedup_in_place].
+    ///
+    /// # Usage
+    /// ```rust
+    /// # use serde_json::json;
+    /// # use serde_json_path::JsonPath;
+    /// # fn main() -> Result<(), serde_json_path::ParseError> {
+    /// let value = json!({"foo": ["bar", "baz"]});
+    /// let path = JsonPath::parse("$.foo[0, 0, 1, 1]")?;
+    /// let nodes = path.query_located(&value);
+    /// assert_eq!(4, nodes.len());
+    /// let nodes = path.query_located(&value).dedup();
+    /// assert_eq!(2, nodes.len());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn dedup(mut self) -> Self {
+        self.dedup_in_place();
+        self
+    }
+
+    /// Deduplicate a [`LocatedNodeList`] _in-place_
+    ///
+    /// See also, [`dedup`][LocatedNodeList::dedup].
+    ///
+    /// # Usage
+    /// ```rust
+    /// # use serde_json::json;
+    /// # use serde_json_path::JsonPath;
+    /// # fn main() -> Result<(), serde_json_path::ParseError> {
+    /// let value = json!({"foo": ["bar", "baz"]});
+    /// let path = JsonPath::parse("$.foo[0, 0, 1, 1]")?;
+    /// let mut nodes = path.query_located(&value);
+    /// assert_eq!(4, nodes.len());
+    /// nodes.dedup_in_place();
+    /// assert_eq!(2, nodes.len());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn dedup_in_place(&mut self) {
+        // This unwrap should be safe, since the paths corresponding to
+        // a query against a Value will always be ordered.
+        self.0
+            .sort_unstable_by(|a, b| a.loc.partial_cmp(&b.loc).unwrap());
```

Comment:
If the paths are already ordered, maybe we can call `dedup` directly?

Reply:
I can see that you may want to mean that, although the elements can be unordered, the result of the same segment are grouped.

But how can we ensure that the unstable sort won't compare two elements of different group (in descendant query)?

Reply:
For example:

```
fn main() {
    let value = json!({
      "o": {"k": [{"j": 1, "k": 2}]},
      "a": [5, 3, [{"j": 4}, {"k": 6}]]
    });
    let path = JsonPath::parse(r#"$["a", "o"]..[0, 1]"#).unwrap();
    let nodes = path.query_located(&value);
    let nodes = path.query_located(&value).dedup();
}

// ["$['a'][0]", "$['a'][1]", "$['a'][2][0]", "$['a'][2][1]", "$['o']['k'][0]"]
```

If the sorter compare `"$['a'][0]"` with `"$['o']['k'][0]"`, it would cause panic.

Reply:
In the other direction, we may sort `PathElement` with names always in the front of index, to make it a total order. Is there some downside I miss?